duplex_http_call.go

```diff
@@ -301,3 +326,81 @@ func cloneURL(oldURL *url.URL) *url.URL {
 	}
 	return newURL
 }
+
+// payloadCloser is a ReadCloser that wraps a bytes.Buffer. It's used to
+// implement the request body for unary calls. To safely reuse the buffer
+// call Wait after the response is received to ensure the payload has been
+// fully read.
+type payloadCloser struct {
+	wait sync.WaitGroup
+
+	mu     sync.Mutex
+	buf    *bytes.Buffer
+	readN  int
+	isDone bool // true if the payload has been fully read
+}
+
+func newPayloadCloser(buf *bytes.Buffer) *payloadCloser {
+	payload := &payloadCloser{
+		buf: buf,
+	}
+	payload.wait.Add(1) // Wait until complete is called
+	return payload
+}
+
+// Read implements io.Reader.
+func (p *payloadCloser) Read(dst []byte) (readN int, err error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.buf == nil {
+		return 0, io.EOF
+	}
+	src := p.buf.Bytes()
+	readN = copy(dst, src[p.readN:])
+	p.readN += readN
+	if readN == 0 && p.readN == len(src) {
+		p.completeWithLock()
+		err = io.EOF
+	}
+	return readN, err
+}
+
+// Close implements io.Closer. It signals that the payload has been fully read.
+func (p *payloadCloser) Close() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.completeWithLock()
+	return nil
+}
+
+// Rewind resets the payload to the beginning. It returns false if the buffer
+// has been discarded from a previous call to Wait.
+func (p *payloadCloser) Rewind() bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.buf == nil {
+		return false
+	}
+	p.readN = 0
+	if p.isDone {
+		p.isDone = false
+		p.wait.Add(1)
+	}
+	return true
+}
+
+// Wait blocks until the payload has been fully read.
+func (p *payloadCloser) Wait() {
+	p.wait.Wait()
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.buf = nil // Discard the buffer
```

Comment:
I think this will defeat the `GetBody` improvement. It seems likely the HTTP framework could rain the entire body (particularly for small bodies) and/or call `Close` and still need to be able to rewind the buffer and send again.

Reply:
`GetBody` shouldn't be called after we've received a response. Draining or calling `Close` will allow `Send` to return and the buffer to be recycled but `GetBody` can call `Rewind()` to reset as many times as needed (5x I think max). The number of `Close` calls or drains isn't counted.